src/ProtocolV3TestBase.sol

```diff
@@ -184,68 +225,202 @@ contract ProtocolV3TestBase is CommonTestBase {
+    uint256 minThresholdAmount = 10 ** config.decimals > 1e8 ? 10 ** config.decimals - 1e8 : 1;
+
+    vm.expectRevert(bytes("36")); // COLLATERAL_CANNOT_COVER_NEW_BORROW
+    pool.borrow(config.underlying, minThresholdAmount, stable ? 1 : 2, 0, borrower);
+
+    vm.stopPrank();
   }
 
-  /**
-   * @dev returns a "good" collateral in the list that cannot be borrowed in stable mode
-   */
-  function _getGoodCollateral(
+  function _e2eTestBorrowRepayWithdraw(
     IPool pool,
-    ReserveConfig[] memory configs,
-    uint256 minSupplyCapDollarMargin
-  ) private view returns (ReserveConfig memory config) {
-    for (uint256 i = 0; i < configs.length; i++) {
-      if (
-        // not frozen etc
-        _includeInE2e(configs[i]) &&
-        // usable as collateral
-        configs[i].usageAsCollateralEnabled &&
-        // not stable borrowable as this makes testing stable borrowing unnecessary hard to reason about
-        !configs[i].stableBorrowRateEnabled &&
-        // LTV is higher than 15%
-        configs[i].ltv > 1500 &&
-        // supply cap not yet reached
-        ((configs[i].supplyCap * 10 ** configs[i].decimals) >
-          IERC20(configs[i].aToken).totalSupply()) &&
-        (// supply cap margin big enough
-        (configs[i].supplyCap * 10 ** configs[i].decimals) -
-          IERC20(configs[i].aToken).totalSupply() >
-          _getTokenAmountByDollarValue(pool, configs[i], minSupplyCapDollarMargin))
-      ) return configs[i];
+    address borrower,
+    ReserveConfig memory collateralConfig,
+    ReserveConfig memory borrowConfig,
+    uint256 amount,
+    bool stable
+  ) internal {
+    if (stable && !borrowConfig.stableBorrowRateEnabled) {
+      console.log('Skip: %s, stable borrow not enabled', borrowConfig.symbol);
+      return;
+    }
+
+    address debtToken = stable ? borrowConfig.stableDebtToken : borrowConfig.variableDebtToken;
+
+    // Step 1: Borrow against collateral
+
+    this._borrow(borrowConfig, pool, borrower, amount, stable);
+
+    // Step 2: Warp to increase interest in system
+
+    vm.warp(block.timestamp + 60 seconds);
+
+    // Step 3: Repay original borrow amount, without accrued interest,
+    //         assert updated state of borrow reserve
+
+    DataTypes.ReserveData memory beforeReserve = pool.getReserveData(borrowConfig.underlying);
+    _repay(borrowConfig, pool, borrower, amount, stable);
+    DataTypes.ReserveData memory afterReserve = pool.getReserveData(borrowConfig.underlying);
+
+    _assertReserveChange(beforeReserve, afterReserve, int256(amount), 60 seconds);
+
+    // Step 4: Try to withdraw all collateral, demonstrate it's not possible without paying back
+    //         accrued debt
+
+    uint256 totalCollateral = IERC20(collateralConfig.aToken).balanceOf(borrower);
+    uint256 remainingDebt   = IERC20(debtToken).balanceOf(borrower);
+
+    // Handle edge case for DAI LTV at 0.01% causing rounding errors here, preventing failure.
+    if (collateralConfig.ltv > 1) {
```

Comment:
Remove the mention of DAI and the ltv should not be specific to the DAI market. I would actually set a lower bound on selecting whether something is "good collateral" in the `_includeCollateralAssetInE2e()` function. That function should skip the DAI market in Spark because it is effectively not collateral. Something like `ltv > 5%` or something to ensure reasonable numbers.

Reply:
Updated, slight correction on what we spoke on:

This function prevents the check of the full collateral amount not being able to be withdrawn because of interest generated on the borrow asset. In the case of low LTV collaterals (e.g., DAI) the amount of asset that can be borrowed is so low that the interest accrued doesn't change the LTV. This condition is still good though because it will test DAI as any other collateral type.